### PR TITLE
Add user profile customizations

### PR DESCRIPTION
**Pull request by smilesthesad**
# Related Issues
Resolves 
<https://github.com/uwcsc/codeybot/issues/140>
<https://github.com/uwcsc/codeybot/issues/141>
<https://github.com/uwcsc/codeybot/issues/142>
<https://github.com/uwcsc/codeybot/issues/143>
<https://github.com/uwcsc/codeybot/issues/144>
<https://github.com/uwcsc/codeybot/issues/145>
<https://github.com/uwcsc/codeybot/issues/146>
<https://github.com/uwcsc/codeybot/issues/147>
<https://github.com/uwcsc/codeybot/issues/148>

# Summary of Changes 
Created userprofiles, where user can set their about me, birth date, preferred name, preferred pronouns, term, year, faculty, program and specialization, which is displayed in a styled embed message when you do .about @someone.

image

# Steps to Reproduce
List the steps needed for the reviewer to produce your PR changes.
- Do .setprofile 
- Follow the instructions, set any one of the customizations
 - Test invalid inputs
 - Test valid inputs
 - Do .about @your_own_name and view your profile

### CODE REVIEW

File: src/commands/profile/profileAbout.ts
Comment (marko-polo-cheno) [inline src/commands/profile/profileAbout.ts:39]: gimme some comments plz

File: src/commands/profile/profileAbout.ts
Comment (marko-polo-cheno) [inline src/commands/profile/profileAbout.ts:47]: interesting idea :D

File: src/components/profile.ts
Comment (alexz429) [inline src/components/profile.ts:52]: I think its spelled february no? Or maybe add both.

File: src/components/db.ts
Comment (marko-polo-cheno) [inline src/components/db.ts:112]: is there a current date thing?

File: src/commands/profile/profileAbout.ts
Comment (alexz429) [inline src/commands/profile/profileAbout.ts:23]: For this (and all future) functionality, try to name commands in the form of .[group][subcommand], like how the interviewer commands are set up.

It helps make sure no future commands overlap on triggers.

Additionally, consider adding a default command to the .profile command, so that the command group has a default behaviour. Something such as how .interviewers is identical to .interviewerslist

File: src/commands/profile/profileSetSpecific.ts
Comment (alexz429) [inline src/commands/profile/profileSetSpecific.ts:11]: see above, go with convention

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:32]: hmm maybe we shouldn't restrict it to just 2000-2030
and use curr year +1 as a max
and u never know if a 1999 grad comes on disc

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:30]: add on alumni

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:90]: lint

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:94]: use enum or smth rather than magical string?

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:99]: all of these can probably be shorter bc u can lowercase and caps first letter on numbers too

File: src/components/profile.ts
Comment (alexz429) [inline src/components/profile.ts:138]: does the IFNULL actually do anything? Even if description is nothing (or NULL), the fact that there is a ' ' around it means that the param is never NULL.

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:137]: dang this is clean :O

Comment (alexz429) [inline src/components/profile.ts:137]: From what I can tell, this loop isn't actually a loop: data only ever has one item? 

If that's true, note that somewhere with a comment, or (if possible) use another way to extract the unknown key-val pair from data.

File: src/components/profile.ts
Comment (alexz429) [inline src/components/profile.ts:147]: note this one somewhere, too

File: src/components/profile.ts
Comment (alexz429) [inline src/components/profile.ts:151]: and this, I guess.
or frankly, at the start of the file, note somewhere that data only has one key-val pair

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:143]: need comments, everywhere XD
this part is not trivial

**Review by marko-polo-cheno — CHANGES_REQUESTED** [review]
solid stuff 👍

File: src/components/profile.ts
Comment (alexz429) [inline src/components/profile.ts:33]: this is a finicky upper bound.

although 2030 does seem far off, I'm graduating in 2025, which means the year after that there's going to be people with graduation dates 2031.

**Review by alexz429 — CHANGES_REQUESTED** [review]

File: src/components/profile.ts
Comment (alexz429) [reply src/components/profile.ts:32]: curr year + 6 probs, since 2022+1 = 2023, therefore I can't even put my year

File: src/components/db.ts
Comment (smilesthesad) [reply src/components/db.ts:112]: Yup, I'll use that instead since this is user-facing so people probably don't care about the exact timestamp

File: src/components/profile.ts
Comment (smilesthesad) [reply src/components/profile.ts:137]: <3

File: src/commands/profile/profileAbout.ts
Comment (smilesthesad) [reply src/commands/profile/profileAbout.ts:23]: A little confused over the second part of this, .interviewers is an alias for .interviewslist. Do you want me to rename profileAbout to profile and then have the name be profile with alias profile-about like so?

image

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:30]: wait still no alumni or prof choice here

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:69]: this must only be your birth month . Must be ...

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:33]: curr_year + 7

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:131]: why not 1 < x and x < 100 instead of valid_years.push(i)

**Review by marko-polo-cheno — CHANGES_REQUESTED** [review]

File: src/commands/profile/profileAbout.ts
Comment (marko-polo-cheno) [reply src/commands/profile/profileAbout.ts]: so basically, 

image

every single class begins with [group]
and there is an interviewers.ts which is exactly the list function (there is no interviewerList.ts

in ur case, since profiles is ur group name, every command should start with ``profiles`` but probably ``profile`` instead.
there should be a profiles.ts that is exactly profilesCommand (what u have) and aliases are about-profile, profile abt, etc

**Review by marko-polo-cheno — APPROVED** [review]

File: src/components/profile.ts
Comment (marko-polo-cheno) [inline src/components/profile.ts:65]: only be your month **of birth**
but eh I guess it's fine as is

**Review by marko-polo-cheno — APPROVED** [review]
usually screenshots would be cool but i peeked at test channel, looks great!!!

**Comment by marko-polo-cheno** [comment]
let's wait for @alexz429 to approve of his requested changes

File: src/components/profile.ts
Comment (smilesthesad) [reply src/components/profile.ts:65]: I can change that when I get home, will merge once alex approves

**Pull request merged**
